packages/backend/server/src/core/workspaces/resolvers/page.ts

```diff
@@ -45,8 +55,133 @@ class WorkspacePage implements Partial<PrismaWorkspacePage> {
   public!: boolean;
 }
 
+@InputType()
+class GrantDocUserRolesInput {
+  @Field(() => String)
+  docId!: string;
+
+  @Field(() => String)
+  workspaceId!: string;
+
+  @Field(() => DocRole)
+  role!: DocRole;
+
+  @Field(() => [String])
+  userIds!: string[];
+}
+
+@InputType()
+class PageGrantedUsersInput {
+  @Field(() => Int)
+  first!: number;
+
+  @Field(() => Int)
+  offset!: number;
+
+  @Field(() => String, { description: 'Cursor', nullable: true })
+  after?: string;
+
+  @Field(() => String, { description: 'Cursor', nullable: true })
+  before?: string;
+}
+
+@ObjectType()
+class GrantedDocUserType {
+  @Field(() => UserType)
+  user!: UserType;
+
+  @Field(() => DocRole)
+  role!: DocRole;
+}
+
+@ObjectType()
+class PageInfo {
+  @Field(() => String, { nullable: true })
+  startCursor?: string;
+
+  @Field(() => String, { nullable: true })
+  endCursor?: string;
+
+  @Field(() => Boolean)
+  hasNextPage!: boolean;
+
+  @Field(() => Boolean)
+  hasPreviousPage!: boolean;
+}
+
+@ObjectType()
+class GrantedDocUserEdge {
+  @Field(() => GrantedDocUserType)
+  user!: GrantedDocUserType;
+
+  @Field(() => String)
+  cursor!: string;
+}
+
+@ObjectType()
+class GrantedDocUsersConnection {
+  @Field(() => Int)
+  totalCount!: number;
+
+  @Field(() => [GrantedDocUserEdge])
+  edges!: GrantedDocUserEdge[];
+
+  @Field(() => PageInfo)
+  pageInfo!: PageInfo;
+}
+
+@ObjectType()
+export class RolePermissions
+  extends WorkspacePermissions
+  implements PermissionsList
+{
+  @Field()
+  Doc_Read!: boolean;
+  @Field()
+  Doc_Copy!: boolean;
+  @Field()
+  Doc_Properties_Read!: boolean;
+  @Field()
+  Doc_Users_Read!: boolean;
+  @Field()
+  Doc_Duplicate!: boolean;
+  @Field()
+  Doc_Trash!: boolean;
+  @Field()
+  Doc_Restore!: boolean;
+  @Field()
+  Doc_Delete!: boolean;
+  @Field()
+  Doc_Properties_Update!: boolean;
+  @Field()
+  Doc_Update!: boolean;
+  @Field()
+  Doc_Publish!: boolean;
+  @Field()
+  Doc_Users_Manage!: boolean;
+  @Field()
+  Doc_TransferOwner!: boolean;
+}
+
+@ObjectType()
+class DocType {
+  @Field(() => String)
+  id!: string;
+
+  @Field(() => Boolean)
+  public!: boolean;
+
+  @Field(() => DocRole)
+  role!: DocRole;
+
+  @Field(() => RolePermissions)
+  permissions!: RolePermissions;
```

Comment:
this is strange. Workspace permissions will be returned in doc type. If separate them, we could even make code cleaner without 'Doc_' prefix